File: sdk/python/shark_auth/webhooks.py
"""Webhooks admin client + HMAC-SHA256 signature verification helper.

Endpoints under ``/api/v1/admin/webhooks/`` (verified in
``internal/api/router.go`` lines 510-522):

- ``POST   /``                                — create
- ``GET    /``                                — list
- ``GET    /events``                          — list event types
- ``GET    /{id}``                            — get
- ``PATCH  /{id}``                            — update
- ``DELETE /{id}``                            — delete
- ``POST   /{id}/test``                       — fire a test delivery
- ``GET    /{id}/deliveries``                 — list past deliveries
- ``POST   /{id}/deliveries/{deliveryId}/replay`` — replay a delivery
"""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


# ---------------------------------------------------------------------------
# Signature verification — module-level helper
# ---------------------------------------------------------------------------


def verify_signature(
    payload: bytes,
    header_signature: str,
    secret: str,
    tolerance_seconds: int = 300,
) -> bool:
    """Verify an HMAC-SHA256 signature on an incoming webhook payload.

    Two header formats are supported:

    1. Stripe-style: ``t=<unix_ts>,v1=<hex_hmac>`` (optionally with extra
       ``vN=...`` segments — only ``v1`` is checked). Signed payload is
       ``f"{ts}.".encode() + payload``. The timestamp is also enforced
       against ``tolerance_seconds`` to thwart replay attacks.
    2. Raw hex digest of HMAC-SHA256(secret, payload). No timestamp enforced.

    Uses :func:`hmac.compare_digest` for timing-safe comparison.

    Parameters
    ----------
    payload:
        Raw request body bytes (DO NOT pass JSON-decoded, must match what the
        server signed byte-for-byte).
    header_signature:
        Value of the signature header sent by SharkAuth.
    secret:
        Shared HMAC secret (the value returned/configured at webhook creation).
    tolerance_seconds:
        Max age of the timestamp in the Stripe-style format. Default 300s.

    Returns
    -------
    bool
        ``True`` if the signature is valid (and, for the timestamped form, the
        timestamp is within tolerance). ``False`` otherwise.

    Example
    -------
    >>> import hmac, hashlib, time
    >>> secret = "whsec_xxx"
    >>> body = b'{"event":"user.created"}'
    >>> ts = int(time.time())
    >>> sig = hmac.new(secret.encode(), f"{ts}.".encode() + body,
    ...                hashlib.sha256).hexdigest()
    >>> header = f"t={ts},v1={sig}"
    >>> assert verify_signature(body, header, secret)
    """
    if not header_signature or not isinstance(header_signature, str):
        return False
    secret_bytes = secret.encode("utf-8") if isinstance(secret, str) else secret

    # Format 1 — comma-separated key=value with t= and v1=.
    if "=" in header_signature and "," in header_signature:
        parts = {}
        for segment in header_signature.split(","):
            segment = segment.strip()
            if "=" in segment:
                k, _, v = segment.partition("=")
                parts[k.strip()] = v.strip()
        ts = parts.get("t")
        v1 = parts.get("v1")
        if ts is None or v1 is None:
            return False
        try:
            ts_int = int(ts)
        except ValueError:
            return False
        # Replay protection.
        if tolerance_seconds > 0 and abs(int(time.time()) - ts_int) > tolerance_seconds:
            return False
        signed = f"{ts}.".encode("utf-8") + payload
        expected = hmac.new(secret_bytes, signed, hashlib.sha256).hexdigest()
        return hmac.compare_digest(expected, v1)

    # Format 2 — raw hex digest (no timestamp; tolerance not enforced).
    expected = hmac.new(secret_bytes, payload, hashlib.sha256).hexdigest()
    # Strip an optional 'sha256=' prefix sometimes used.
    candidate = header_signature.strip()
    if candidate.lower().startswith("sha256="):
        candidate = candidate.split("=", 1)[1]
    return hmac.compare_digest(expected, candidate)


# ---------------------------------------------------------------------------
# Admin client
# ---------------------------------------------------------------------------


class WebhooksClient:
    """Admin client for webhook endpoints.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    admin_api_key:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.

    Example
    -------
    >>> wh = WebhooksClient(base_url="https://auth.example.com",
    ...                     admin_api_key="sk_live_xxx")
    >>> created = wh.register(
    ...     url="https://app.example.com/hooks",
    ...     events=["user.created", "user.deleted"],
    ...     secret="whsec_xxx",
    ... )
    """

    _PREFIX = "/api/v1/admin/webhooks"

    def __init__(
        self,
        base_url: str,
        admin_api_key: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = admin_api_key
        self._session = session or _http.new_session()

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    def register(
        self,
        url: str,
        events: List[str],
        *,
        secret: Optional[str] = None,
        **extra: Any,
    ) -> Dict[str, Any]:
        """Create a webhook subscription.

        Parameters
        ----------
        url:
            HTTPS endpoint that will receive POSTed events.
        events:
            List of event-type strings to subscribe to.
        secret:
            Optional HMAC secret for payload signing. If omitted the server
            may auto-generate one and return it in the response.
        **extra:
            Additional fields forwarded verbatim (e.g. ``description``,
            ``enabled``).
        """
        body: Dict[str, Any] = {"url": url, "events": events}
        if secret is not None:
            body["secret"] = secret
        for k, v in extra.items():
            body[k] = v
        endpoint = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "POST", endpoint, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return resp.json()
        _raise(resp)

    def list(self) -> List[Dict[str, Any]]:
        """List all webhook subscriptions."""
        endpoint = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "GET", endpoint, headers=self._auth())
        if resp.status_code == 200:
            data = resp.json()
            if isinstance(data, dict):
                return data.get("data", [])
            return data
        _raise(resp)

    def get(self, webhook_id: str) -> Dict[str, Any]:
        """Fetch a single webhook subscription."""
        endpoint = f"{self._base}{self._PREFIX}/{webhook_id}"
        resp = _http.request(self._session, "GET", endpoint, headers=self._auth())
        if resp.status_code == 200:
            return resp.json()
        _raise(resp)

    def update(self, webhook_id: str, **fields: Any) -> Dict[str, Any]:
        """Patch a webhook subscription.

        Backend uses ``PATCH /{id}`` (verified in router.go line 517).
        """
        endpoint = f"{self._base}{self._PREFIX}/{webhook_id}"
        resp = _http.request(
            self._session, "PATCH", endpoint, headers=self._auth(), json=fields
        )
        if resp.status_code == 200:
            return resp.json()
        _raise(resp)

    def delete(self, webhook_id: str) -> None:
        """Delete a webhook subscription."""
        endpoint = f"{self._base}{self._PREFIX}/{webhook_id}"
        resp = _http.request(self._session, "DELETE", endpoint, headers=self._auth())
        if resp.status_code in (200, 204):
            return
        _raise(resp)

    # ------------------------------------------------------------------
    # Test + deliveries
    # ------------------------------------------------------------------

    def send_test(
        self,
        webhook_id: str,
        *,
        event_type: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Fire a synthetic test event.

        Backend path: ``POST /{id}/test`` (verified in router.go line 519 —
        note the path is ``/test``, not ``/send-test``).
        """
        endpoint = f"{self._base}{self._PREFIX}/{webhook_id}/test"
        body: Dict[str, Any] = {}
        if event_type is not None:
            body["event_type"] = event_type
        resp = _http.request(
            self._session,
            "POST",
            endpoint,
            headers=self._auth(),
            json=body if body else None,
        )
        if resp.status_code in (200, 201, 202):
            try:
                return resp.json()
            except Exception:
                return {}
        _raise(resp)

    def replay(self, webhook_id: str, delivery_id: str) -> Dict[str, Any]:
        """Replay a previous delivery."""
        endpoint = (
            f"{self._base}{self._PREFIX}/{webhook_id}/deliveries/{delivery_id}/replay"
        )
        resp = _http.request(self._session, "POST", endpoint, headers=self._auth())
        if resp.status_code in (200, 201, 202):
            try:
                return resp.json()
            except Exception:
                return {}
        _raise(resp)

    def list_deliveries(
        self,
        webhook_id: str,
        *,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """List past deliveries for a webhook."""
        endpoint = f"{self._base}{self._PREFIX}/{webhook_id}/deliveries"
        resp = _http.request(
            self._session,
            "GET",
            endpoint,
            headers=self._auth(),
            params={"limit": limit},
        )
        if resp.status_code == 200:
            data = resp.json()
            if isinstance(data, dict):
                return data.get("data", [])
            return data
        _raise(resp)


__all__ = ["WebhooksClient", "verify_signature"]

File: sdk/typescript/src/webhooks.ts
/**
 * Webhooks admin client + HMAC-SHA256 signature verification helper.
 *
 * Endpoints under `/api/v1/admin/webhooks/` (verified in
 * `internal/api/router.go` lines 510-522):
 *
 *  - POST   /                                — create
 *  - GET    /                                — list
 *  - GET    /{id}                            — get
 *  - PATCH  /{id}                            — update
 *  - DELETE /{id}                            — delete
 *  - POST   /{id}/test                       — fire a test delivery
 *  - GET    /{id}/deliveries                 — list past deliveries
 *  - POST   /{id}/deliveries/{deliveryId}/replay — replay a delivery
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Webhook subscription row as returned by the server. */
export interface Webhook {
  id: string;
  url: string;
  events: string[];
  enabled?: boolean;
  description?: string;
  created_at?: string;
  updated_at?: string;
  [key: string]: unknown;
}

/** Webhook delivery record. */
export interface WebhookDelivery {
  id: string;
  webhook_id?: string;
  event_type?: string;
  status?: string;
  status_code?: number;
  attempts?: number;
  delivered_at?: string;
  created_at?: string;
  [key: string]: unknown;
}

/** Input for {@link WebhooksClient.register}. */
export interface RegisterWebhookInput {
  url: string;
  events: string[];
  secret?: string;
  description?: string;
  enabled?: boolean;
  [key: string]: unknown;
}

/** Input for {@link WebhooksClient.update} — partial. */
export type UpdateWebhookInput = Partial<RegisterWebhookInput>;

/** Options for {@link WebhooksClient}. */
export interface WebhooksClientOptions {
  baseUrl: string;
  adminKey: string;
}

// ---------------------------------------------------------------------------
// Module-level signature verification helper
// ---------------------------------------------------------------------------

const HEX_CHARS = "0123456789abcdef";

function bytesToHex(bytes: ArrayBuffer | Uint8Array): string {
  const view = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let out = "";
  for (let i = 0; i < view.length; i++) {
    const b = view[i] ?? 0;
    out += HEX_CHARS[(b >> 4) & 0xf];
    out += HEX_CHARS[b & 0xf];
  }
  return out;
}

function utf8Encode(input: string | Uint8Array): Uint8Array {
  if (input instanceof Uint8Array) return input;
  return new TextEncoder().encode(input);
}

function concatBytes(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

/** Constant-time hex string comparison. */
function timingSafeEqualHex(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

async function hmacSha256Hex(secret: Uint8Array, data: Uint8Array): Promise<string> {
  // Copy into fresh ArrayBuffers to satisfy lib.dom BufferSource typing
  // across Node and DOM lib variants.
  const secretBuf = secret.buffer.slice(
    secret.byteOffset,
    secret.byteOffset + secret.byteLength,
  );
  const dataBuf = data.buffer.slice(
    data.byteOffset,
    data.byteOffset + data.byteLength,
  );
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const subtle: any = (globalThis as any).crypto?.subtle;
  if (!subtle) throw new Error("Web Crypto subtle API not available");
  const key = await subtle.importKey(
    "raw",
    secretBuf,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sigBuf = (await subtle.sign("HMAC", key, dataBuf)) as ArrayBuffer;
  return bytesToHex(sigBuf);
}

/**
 * Verify an HMAC-SHA256 signature on an incoming webhook payload.
 *
 * Two header formats are supported:
 *
 *   1. Stripe-style: `t=<unix_ts>,v1=<hex_hmac>` (extra `vN=` segments
 *      ignored). Signed payload is `${ts}.<payload>`. Timestamp is
 *      enforced against `toleranceSeconds` to thwart replay attacks.
 *   2. Raw hex digest of HMAC-SHA256(secret, payload). Optional
 *      `sha256=` prefix is stripped. No timestamp enforced.
 *
 * Uses Web Crypto (`crypto.subtle`) — works in browsers and Node 18+.
 *
 * @param payload          Raw request body bytes or string.
 * @param headerSignature  Signature header value sent by SharkAuth.
 * @param secret           Shared HMAC secret.
 * @param toleranceSeconds Max age of the timestamp (Stripe-style only). Default 300.
 * @returns `true` if valid, `false` otherwise.
 */
export async function verifySignature(
  payload: Uint8Array | string,
  headerSignature: string,
  secret: string,
  toleranceSeconds: number = 300,
): Promise<boolean> {
  if (!headerSignature || typeof headerSignature !== "string") return false;
  const secretBytes = utf8Encode(secret);
  const payloadBytes = utf8Encode(payload);

  // Format 1 — comma-separated key=value pairs.
  if (headerSignature.includes("=") && headerSignature.includes(",")) {
    const parts: Record<string, string> = {};
    for (const seg of headerSignature.split(",")) {
      const trimmed = seg.trim();
      const eq = trimmed.indexOf("=");
      if (eq > 0) {
        parts[trimmed.slice(0, eq).trim()] = trimmed.slice(eq + 1).trim();
      }
    }
    const ts = parts["t"];
    const v1 = parts["v1"];
    if (ts === undefined || v1 === undefined) return false;
    const tsInt = Number.parseInt(ts, 10);
    if (!Number.isFinite(tsInt)) return false;
    if (toleranceSeconds > 0) {
      const nowSec = Math.floor(Date.now() / 1000);
      if (Math.abs(nowSec - tsInt) > toleranceSeconds) return false;
    }
    const signed = concatBytes(utf8Encode(`${ts}.`), payloadBytes);
    const expected = await hmacSha256Hex(secretBytes, signed);
    return timingSafeEqualHex(expected, v1.toLowerCase());
  }

  // Format 2 — raw hex digest, optionally prefixed with "sha256=".
  let candidate = headerSignature.trim();
  if (candidate.toLowerCase().startsWith("sha256=")) {
    candidate = candidate.slice("sha256=".length);
  }
  const expected = await hmacSha256Hex(secretBytes, payloadBytes);
  return timingSafeEqualHex(expected, candidate.toLowerCase());
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

const PREFIX = "/api/v1/admin/webhooks";

/** Admin client for webhook subscriptions. */
export class WebhooksClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: WebhooksClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string } | string;
        message?: string;
      };
      if (typeof body.error === "object" && body.error) {
        if (body.error.code) code = body.error.code;
        if (body.error.message) message = body.error.message;
      } else if (typeof body.error === "string") {
        message = body.error;
      } else if (body.message) {
        message = body.message;
      }
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  // --------------------------------------------------------------- CRUD

  /** Create a webhook subscription. */
  async register(input: RegisterWebhookInput): Promise<Webhook> {
    const url = `${this._base}${PREFIX}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(input),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    const raw = resp.json<Webhook | { data: Webhook }>();
    if (raw && typeof raw === "object" && "data" in raw) return (raw as { data: Webhook }).data;
    return raw as Webhook;
  }

  /** List all webhook subscriptions. */
  async list(): Promise<Webhook[]> {
    const url = `${this._base}${PREFIX}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<Webhook[] | { data?: Webhook[] }>();
    if (Array.isArray(body)) return body;
    return body?.data ?? [];
  }

  /** Fetch a single webhook subscription. */
  async get(webhookId: string): Promise<Webhook> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(webhookId)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const raw = resp.json<Webhook | { data: Webhook }>();
    if (raw && typeof raw === "object" && "data" in raw) return (raw as { data: Webhook }).data;
    return raw as Webhook;
  }

  /** Patch a webhook subscription (PATCH /{id}). */
  async update(webhookId: string, fields: UpdateWebhookInput): Promise<Webhook> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(webhookId)}`;
    const resp = await httpRequest(url, {
      method: "PATCH",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(fields),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const raw = resp.json<Webhook | { data: Webhook }>();
    if (raw && typeof raw === "object" && "data" in raw) return (raw as { data: Webhook }).data;
    return raw as Webhook;
  }

  /** Delete a webhook subscription. */
  async delete(webhookId: string): Promise<void> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(webhookId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // --------------------------------------------------------- Test + deliveries

  /** Fire a synthetic test event (POST /{id}/test). */
  async sendTest(
    webhookId: string,
    opts?: { eventType?: string },
  ): Promise<Record<string, unknown>> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(webhookId)}/test`;
    const body: Record<string, unknown> = {};
    if (opts?.eventType !== undefined) body.event_type = opts.eventType;
    const hasBody = Object.keys(body).length > 0;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: hasBody
        ? { ...this._auth(), "Content-Type": "application/json" }
        : this._auth(),
      body: hasBody ? JSON.stringify(body) : undefined,
    });
    if (![200, 201, 202].includes(resp.status))
      await this._throw(resp.status, resp.text);
    try {
      return resp.json<Record<string, unknown>>();
    } catch {
      return {};
    }
  }

  /** Replay a previous delivery. */
  async replay(webhookId: string, deliveryId: string): Promise<Record<string, unknown>> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(
      webhookId,
    )}/deliveries/${encodeURIComponent(deliveryId)}/replay`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (![200, 201, 202].includes(resp.status))
      await this._throw(resp.status, resp.text);
    try {
      return resp.json<Record<string, unknown>>();
    } catch {
      return {};
    }
  }

  /** List past deliveries for a webhook. */
  async listDeliveries(
    webhookId: string,
    opts?: { limit?: number },
  ): Promise<WebhookDelivery[]> {
    const limit = opts?.limit ?? 50;
    const qs = new URLSearchParams({ limit: String(limit) });
    const url = `${this._base}${PREFIX}/${encodeURIComponent(
      webhookId,
    )}/deliveries?${qs.toString()}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<WebhookDelivery[] | { data?: WebhookDelivery[] }>();
    if (Array.isArray(body)) return body;
    return body?.data ?? [];
  }
}
